x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/SearchableSnapshotAction.java

```diff
@@ -112,57 +151,124 @@ boolean isForceMergeIndex() {
-                            "snapshot. Skipping this action", SearchableSnapshotAction.NAME, index.getName(),
-                        LifecycleSettings.LIFECYCLE_NAME_SETTING.get(indexMetadata.getSettings()));
-                    return true;
+                    // The index is already a searchable snapshot, let's see if the repository matches
+                    // TODO: move the searchable snapshot settings into x-pack
+                    //  core in the future, so the Settings can be used instead
+                    //  of strings here
+                    String repo = indexMetadata.getSettings().get("index.store.snapshot.repository_name");
+                    if (this.snapshotRepository.equals(repo) == false) {
+                        // Okay, different repo, we need to go ahead with the searchable snapshot
+                        logger.debug("[{}] action is configured for index [{}] in policy [{}] which is already mounted as a searchable " +
+                                "snapshot, but with a different repository (existing: [{}] vs new: [{}]), a new snapshot and " +
+                                "index will be created",
+                            SearchableSnapshotAction.NAME, index.getName(), policyName, repo, this.snapshotRepository);
+                        return false;
+                    }
+
+                    // Check to the storage type to see if we need to convert between full <-> partial
+                    boolean partial = indexMetadata.getSettings().getAsBoolean("index.store.snapshot.partial", false);
+                    MountSearchableSnapshotRequest.Storage existingType =
+                        partial ? MountSearchableSnapshotRequest.Storage.SHARED_CACHE : MountSearchableSnapshotRequest.Storage.FULL_COPY;
+                    MountSearchableSnapshotRequest.Storage type = getConcreteStorageType(preActionBranchingKey);
+                    if (existingType == type) {
+                        logger.debug("[{}] action is configured for index [{}] in policy [{}] which is already mounted " +
+                            "as a searchable snapshot with the same repository [{}] and storage type [{}], skipping this action",
+                            SearchableSnapshotAction.NAME, index.getName(), policyName, repo, type);
+                        return true;
+                    }
+
+                    logger.debug("[{}] action is configured for index [{}] in policy [{}] which is already mounted " +
+                        "as a searchable snapshot, however, the repository ([{}] vs [{}]) or storage type ([{}] vs [{}]) " +
+                        "differs, so a new index will be created",
+                        SearchableSnapshotAction.NAME, index.getName(), policyName, repo, this.snapshotRepository, existingType, type);
+                    // Perform the searchable snapshot
+                    return false;
```

Comment:
Maybe I'm misunderstanding the code here but I believe this statement is not true? We are performing the action because the storage type is different not because the repos are different? (we already handled the case where the repos are different [above](https://github.com/elastic/elasticsearch/pull/68714/files#diff-ce8f48edd70fb6cf0b3c24269aac86dae18d5bc3c0382f0d7669c8a446c4fa45R161) )

Reply:
Yeah, it can be changed to just print the current repository since it's only checked the difference in storage type :+1: